File: client/src/test/scala/stasis/test/specs/unit/client/encryption/secrets/DeviceMetadataSecretSpec.scala
package stasis.test.specs.unit.client.encryption.secrets

import org.apache.pekko.actor.ActorSystem
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import stasis.client.encryption.secrets.DeviceMetadataSecret
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.EncodingHelpers

class DeviceMetadataSecretSpec extends AsyncUnitSpec with SecretsConfig with EncodingHelpers {
  "A DeviceMetadataSecret" should "support encryption" in {
    Source
      .single(ByteString(plaintextData))
      .via(metadataSecret.encryption)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualEncryptedData =>
        actualEncryptedData should be(encryptedData.decodeFromBase64)
      }
  }

  it should "support decryption" in {
    Source
      .single(encryptedData.decodeFromBase64)
      .via(metadataSecret.decryption)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualPlaintextData =>
        actualPlaintextData should be(ByteString(plaintextData))
      }
  }

  it should "not render its content via toString" in {
    metadataSecret.toString should be(s"Secret(${metadataSecret.getClass.getName})")
  }

  private implicit val system: ActorSystem = ActorSystem(name = "DeviceMetadataSecretSpec")

  private val encryptionIv = "kUuYeWjrwqnA93zYCXn2ZC3Pr5Y4srYEcgrR3jP5KtM="
  private val encryptionKey = "QBqEu8Kh6iFGpbgYUWADXRfkVa6wUy5w"

  private val plaintextData = "some-plaintext-data"
  private val encryptedData = "coKEpIHZVHZtPcYuojvMPcOD8S6a2sH4Xg0gPX8UTKz75Ts="

  private val metadataSecret = DeviceMetadataSecret(
    iv = encryptionIv.decodeFromBase64,
    key = encryptionKey.decodeFromBase64
  )

}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/encryption/secrets/DeviceMetadataSecretSpec.kt
package stasis.test.client_android.lib.encryption.secrets

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.Buffer
import okio.ByteString.Companion.decodeBase64
import okio.Source
import okio.buffer
import stasis.client_android.lib.encryption.secrets.DeviceMetadataSecret

class DeviceMetadataSecretSpec : WordSpec({
    "A DeviceMetadataSecret" should {
        val encryptionIv = "kUuYeWjrwqnA93zYCXn2ZC3Pr5Y4srYEcgrR3jP5KtM="
        val encryptionKey = "QBqEu8Kh6iFGpbgYUWADXRfkVa6wUy5w"

        val plaintextData = "some-plaintext-data"
        val encryptedData = "coKEpIHZVHZtPcYuojvMPcOD8S6a2sH4Xg0gPX8UTKz75Ts="

        val metadataSecret = DeviceMetadataSecret(
            iv = encryptionIv.decodeBase64()!!,
            key = encryptionKey.decodeBase64()!!
        )

        "support encryption" {
            val actualEncryptedData = metadataSecret
                .encryption(Buffer().writeUtf8(plaintextData) as Source)
                .buffer()
                .readByteString()

            actualEncryptedData shouldBe (encryptedData.decodeBase64()!!)
        }

        "support decryption" {
            val actualPlaintextData = metadataSecret
                .decryption(Buffer().write(encryptedData.decodeBase64()!!))
                .buffer()
                .readUtf8()

            actualPlaintextData shouldBe (plaintextData)
        }

        "not render its content via toString" {
            metadataSecret.toString() shouldBe ("Secret(${metadataSecret.javaClass.name})")
        }
    }
})
